components/hab/src/license.rs

```diff
@@ -0,0 +1,163 @@
+    pub file_format: String,
+}
+
+impl LicenseData {
+    pub fn new() -> Self {
+        LicenseData { date_accepted: Utc::now(),
+                      accepting_product: String::from("hab"),
+                      accepting_product_version: super::VERSION.to_string(),
+                      user: get_current_username(),
+                      file_format: String::from(LICENSE_FILE_FORMAT_VERSION), }
+    }
+}
+
+pub fn check_for_license_acceptance_and_prompt(ui: &mut UI) -> Result<()> {
+    if license_exists() || env_var_present()? {
+        return Ok(());
+    }
+
+    ui.heading("+---------------------------------------------+")?;
+    ui.heading("            Chef License Acceptance")?;
+    ui.br()?;
+    ui.info("Before you can continue, 1 product license must be accepted.")?;
+    ui.info("View the license at https://www.chef.io/end-user-license-agreement")?;
+    ui.br()?;
+    ui.info("License that needs accepting:")?;
+    ui.br()?;
+    ui.info("  * Habitat")?;
+    ui.br()?;
+
+    if ui.prompt_yes_no("Do you accept the 1 product license?", Some(false))? {
+        accept_license(ui)
+    } else {
+        ui.br()?;
+        ui.info("If you do not accept this license you will not be able to use Chef products.")?;
+        ui.br()?;
+
+        if ui.prompt_yes_no("Do you accept the 1 product license?", Some(false))? {
+            accept_license(ui)
+        } else {
+            ui.br()?;
+            ui.heading("+---------------------------------------------+")?;
+            Err(Error::LicenseNotAccepted)
+        }
+    }
+}
+
+pub fn accept_license(ui: &mut UI) -> Result<()> {
+    if license_exists() {
+        ui.info("You have already accepted the license.")?;
+        return Ok(());
+    }
+
+    ui.br()?;
+    ui.info("Accepting 1 product license...")?;
+
+    write_license_file()?;
+
+    ui.status(ui::Status::Custom(ui::Glyph::CheckMark, String::from("")),
+              "1 product license accepted.")?;
+    ui.br()?;
+    ui.heading("+---------------------------------------------+")?;
+
+    Ok(())
+}
+
+fn license_path() -> PathBuf {
+    let hab_dir = if am_i_root() {
+        PathBuf::from(&*FS_ROOT_PATH).join("hab")
+    } else if let Some(home) = dirs::home_dir() {
+        home.join(".hab")
+    } else {
+        panic!("No home directory available. Unable to find a suitable place to write a license \
+                file.");
+    };
+
+    hab_dir.join("accepted-licenses")
+}
+
+fn env_var_present() -> Result<bool> {
+    match env::var(LICENSE_ACCEPT_ENVVAR) {
+        Ok(val) => {
+            if &val == "accept" {
+                write_license_file()?;
+                Ok(true)
+            } else if &val == "accept-no-persist" {
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+        Err(_) => Ok(false),
+    }
+}
+
+fn write_license_file() -> Result<()> {
+    let license = LicenseData::new();
+    let content = serde_yaml::to_string(&license)?;
```

Comment:
Honestly I'm not crazy about changing from code that handles an error gracefully to one that panics.  I don't think I understand how that change would improve this code.

Reply:
I don't think anything is hurt by leaving the `?` here.

However, there's absolutely nothing outside of our control that would cause this to fail. Indeed, since we've just automatically derived `Serialize` on `LicenseData`, Serde itself would need to be broken for this line to throw an error (I took a similar approach [here](https://github.com/habitat-sh/habitat/pull/6344/commits/70c74040b456cdaec43f618370adf4f6c6efd956#diff-83ad8fbb200c6854b003010c49e34dc3R58)). All the other error cases in this function deal with things outside our control (does the user have permission to create a directory? a file? permission to write to a file?), so using `?` makes sense.

Having done some recent digging through our error handling code, I'm also growing skeptical of our existing approach of simply adding a "wrapper" error case and a `From` implementation to deal with errors from our various dependencies:

```rust
impl From<serde_yaml::Error> for Error {
    fn from(err: serde_yaml::Error) -> Error { Error::YamlError(err) }
}
```
Because of how `?` works, basically any YAML error anywhere in this crate will now be turned into a "hab YAML error", without any further context. This is currently the only case right now, so it's no big deal, but it makes it incredibly easy for this to creep into any other YAML-related functions we might add in the future. If we had a more principled approach to adding errors, this particular error (if we wanted to even keep it an error, depending on whether my argument above holds any water) might become a `SerializeLicenseError`, or something similar. It could still contain the original `serde_yaml::Error` instance for original cause information, but now holds the added information that this is a very specific kind of YAML error. We would move away from simply "passing errors along" to creating domain-specific, context-rich errors.

With that kind of error approach, you'd end up writing something like:

```rust
let content = serde_yaml::to_string(&license).map_err(SerializeLicenseError)?;
```

instead of:

```rust
let content = serde_yaml::to_string(&license)?;
```
(That's all a roundabout way of saying that if we don't throw the YAML error here, then we also don't need to add the additional error type and impl.)

I know our error situation is going to take a bit of discussion and work to clean up, so I'm not looking for anything to change here, but seeing this brought all these thoughts to mind, which is why I brought it up.

Reply:
That makes sense.  Thanks for elaborating.

I think it makes sense to leave it for now and update it when we revamp errors as a whole.